cmd/recorder.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2020-2021 JackTrip Labs, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mewkiz/flac"
+	"github.com/mewkiz/flac/frame"
+	"github.com/mewkiz/flac/meta"
+	"github.com/mewkiz/pkg/pathutil"
+	"github.com/xthexder/go-jack"
+)
+
+const (
+	// FileDuration is the duration (in seconds) of each audio segment file
+	FileDuration = 10
+	// FileCountLimit is the maximum number of audio segment files kept on disk
+	FileCountLimit = 5
+	// NumRecorderChannels is the number of input channels of the recorder
+	NumRecorderChannels = 2
+	// BitDepth is the bit-resolution used when encoding audio data - changing this involves changes in processBuffer()
+	BitDepth = 16
+	// HLSIndex is the top-level HLS metadata file
+	HLSIndex = "index.m3u8"
+)
+
+var (
+	// AudioFilenames is an in-memory array of filenames used to perform rotation
+	AudioFilenames []string
+	// FrameBuffer is an in-memory buffer of FLAC frames
+	FrameBuffer []frame.Frame
+	// HLSPlaylistHash is a randomly generated hash to uniquely identify playlists on restart of jackd/jacktrip
+	HLSPlaylistHash string
+)
+
+// Recorder listens to audio and records it to disk
+type Recorder struct {
+	// Name will be the JACK client name
+	Name string
+	// JackBufferSize is the buffer size determined at the start of client initiation
+	JackBufferSize int
+	// JackSampleRate is the sample rate determined at the start of client initiation
+	JackSampleRate int
+	// JackClient is the client interacting with jackd
+	JackClient *jack.Client
+	// RecorderPorts are the input ports used to scrape audio
+	RecorderPorts []*jack.Port
+	// ClientLock is a mutex used for client-daemon interactions
+	ClientLock sync.Mutex
+	// RawSamplesChan is the channel were raw audio samples are passed through
+	RawSamplesChan chan [][]jack.AudioSample
+}
+
+// NewRecorder constructs a new instance of Recorder
+// TODO: Should we merge this and make it part of autoconnector?
```

Comment:
Not sure I know the answer, but something to keep in mind is that we may want to reuse this pattern for other things in the (near?) future. For example, having a recorder connected to the broadcast jack ports of every jacktrip client so that we are able to record lossless raw audio (with the added buffering of the broadcast feature) for every musician, independently. This would be extremely valuable for recording, for example.

Reply:
Cool - I'm going to leave it as-is then